techlibs/xilinx/xc7_xcu_brams.txt

```diff
@@ -85,6 +85,7 @@ match $__XILINX_RAMB18_SDP
   min efficiency 5
   shuffle_enable B
   make_transp
+  attribute ram_style block
```

Comment:
My original suggestion would be to:
(a) add support for `attribute` and
(b) use it in a new set of match blocks (if it is indeed possible to have multiple match blocks for the same `bram` entry) as I would think that would be more flexible....

Take for example that you don't want to ignore `min_bits` or`efficiency` (or any other condition) but instead have different thresholds or conditions outright.... do you think that would be a good idea?

Reply:
The thing is, `parse_match` function does not have a way to know if the match name is the same, but with different rules, it parses line by line starting from `match` keyword. If we put two match rules with same name, the memory will ended into a RAMB18E1 all the time. That's why I didn't used that approach.

There are more attributes than `block`, that can be applied to a memory, so I think the flexibility would be more into `memory_bram` pass instead of the match rules. For instance, attribute `block` forces Vivado to use RAMB18 iff memory words are greather than or equal to 2 (therefore, min bits and efficiency are ignored), so for this specific case we could use `min words N` (maybe more than 2 if we don't want to use Vivado thresholds). Quick way would be to copy and rename both `$__XILINX_RAMB18_SDP` match rule and Verilog techmap instantiation, (like ```$__XILINX_RAMB18_SDP_ATTR```) and define there the `min word` constraint, but we'll have redundant RAMB18E1 techmap. We could also manage that case in the `memory_bram` pass, but we need a mechanism to know every supported family attributes if we don't plan to make this Xilinx spcefic. Or, make `memory_bram` recognise different match rules for same memory.

Reply:
> The thing is, parse_match function does not have a way to know if the match name is the same, but with different rules, it parses line by line starting from match keyword. 

If I'm understanding correctly, then it's possible to have multiple `match` blocks with the same "name" (referring to the same `bram` entry) but with different conditions -- right? That's actually what I was hoping for, because you can have multiple entries:

```
match $__XILINX_RAMB18_SDP
  min bits 4096
  min efficiency 5
  shuffle_enable B
  make_transp
  attribute ram_style # No/empty value indicating this attribute must not exist
  or_next_if_better
endmatch

# This match block has no min bits/efficiency conditions, just ram_style="block"
match $__XILINX_RAMB18_SDP
  shuffle_enable B
  make_transp
  attribute ram_style block
  or_next_if_better
endmatch
```
Is that possible, or am I missing something?

> If we put two match rules with same name, the memory will ended into a RAMB18E1 all the time. That's why I didn't used that approach.

Not sure what you mean here -- can you elaborate please?

Reply:
Resolved in 89be614